v0 API structured data compliance (#105)
nautobot/nautobot-plugin-golden-config: docs/navigating-compliance-json.md

```json
{
  "actual": "{\"aaa\": {\"servers\": [\"1.1.1.1\"]}}",
  "intended": "{\"aaa\": {\"servers\": [\"1.1.1.1\"]}}",
  "device": "472c3786-4499-4e34-8d8e-291c4c7f32a9",
  "rule": "87cc003b-9025-472d-bf3c-e5b62bf91987"
}
```

> Notice: The UUID is needed for the device and rule to execute against.  This data can be found in the UI or by querying the proper endpoints, `/plugins/golden-config/compliance-rule/` and `/dcim​/devices​/`.

Once the API call is made the response data provides a quick snapshot.
```json
{
  "id": "be35a701-0cb0-418a-98e7-a82d99db857e",
  "device": "472c3786-4499-4e34-8d8e-291c4c7f32a9",
  "rule": "87cc003b-9025-472d-bf3c-e5b62bf91987",
  "actual": "{\"aaa\": {\"servers\": [\"1.1.1.1\"]}}",
  "intended": "{\"aaa\": {\"servers\": [\"1.1.1.1\"]}}",
  "ordered": true,
  "compliance": true,
  "display": "nautobot-iosv-1 -> iosv-json - aaa-json -> True"
}
```

You can also see the compliance data in the UI once it is created via API.

In the navigation menu: `Plugins -> Configuration Compliance`.

![Example Compliance Run in UI](./img/03-navigating-compliance-json.png)

> 

Reviewer: Can you add an example when it is not compliant?

# Navigating Compliance Using Structured Data

> Warning:
This feature is in early development and is only implemented within the API.  There are multiple caveats that need to be considered and are listed below.

## Caveats
- Must have a `Platform` created specifically for the feature.

> - Both the compliance `feature` and `rule` need to be unique for the JSON `config-type`.

Reviewer: Why is this again? Why not just the rule?

Author: In the ComplianceRule model, the unique_together constraint is feature and platform.

Author: Its just rule.  Will fix.

![Example Compliance Run in UI](./img/03-navigating-compliance-json.png)

With the detailed view:

![Example Compliance Run in UI Detail](./img/04-navigating-compliance-json.png)

Example of a Non-Compliant rule:

![Example Non-Compliant Run in UI Detail](./img/05-navigating-compliance-json.png)

> Note:
All subsequent updates to the compliance rule need to use the `PUT` or `PATCH` API methods and they need to provide the UUID that was created for the specific compliance execution. This UUID is generated by (unique_together = ("device", "rule"))

As soon as a Configuration Compliance object has been created using the API seen above, the `Configuration Overview` will be available, and will show the **diff** action button.

![Example of Diff Icon](./img/06-navigating-compliance-json.png)

Seeing the diff button alone will **only** show up for devices using JSON compliance rules.

> Note: The backend will process and find all compliance rule objects that exist for the device and the create the a temporary backup and intended configurations on the fly. Meaning, when the **diff** action button is clicked all of the processes happens at that time, to render the diff.

> If the page is refreshed the Status and other icons will now appear.

Reviewer: I think this should be removed.

# Navigating Compliance Using Structured Data

> Warning:
This feature is in early development and is only implemented within the API.  There are multiple caveats that need to be considered and are listed below.

## Caveats
- Must have a `Platform` created specifically for the feature.
- Both the compliance `feature` and `rule` need to be unique for the JSON `config-type`.

> - When creating `Compliance Rules` with the config type of JSON, the `config to match` field is not used but must have bogus data in order to save to the DB.

Reviewer: Let's change this to not required on model and validate when cli, it is required.

Author: changed `blank=True`.  Will add further checks.

> - Does **NOT** require Git repositories and for backup and intended.

Reviewer: ```suggestion
- Does **NOT** use Git repositories and for backup and intended.
```

> - Mixing/Matching CLI and JSON compliance rules is **NOT** supported.  A device should only have compliance rules from one or the other.

Reviewer: ```suggestion
- Mixing/Matching CLI and JSON compliance rules is **NOT** supported.  A device should only have compliance rules from one or the other, and it is up to the operator to enforce.
```

## Quick Start Guide

1. To get started with the structured data(JSON) compliance check the first step is to create a platform.

![Example Platform Creation](./img/00-navigating-compliance-json.png)

2. Next, the compliance feature needs to be created.
![Example Feature Creation](./img/01-navigating-compliance-json.png)

3. Link the feature that was just created to a rule definition.
![Example Rule Creation](./img/02-navigating-compliance-json.png)

> > Notice: Bogus text data `none` was added to `Config to Match` which is not used in the current implementation.

Reviewer: Note: update when change to not required.

4. Now that the definitions are created and the rule is mapped to the platform we created the API execution can be performed.

Here is an example of a `POST` to the `/plugins​/golden-config​/config-compliance​/` endpoint.

> ```json

Reviewer: Is this still correct with jsonfield?

Reviewer: @jeffkala can you confirm?

Author: @itdependsnetworks trying to figure out how I get this to work.  Even with jsonfield I need to escape everything for it to work.  Maybe I need to overload the create function on the api serializer?  Trying to do some reading on what options I have.

Author: Still need to send it like this.
```python
data = {
  "actual": json.dumps({"ospf": ["8.8.8.8"]}),
  "intended": json.dumps({"ospf": ["8.8.8.8", "80.80.80.80"]}),
  "device": "472c3786-4499-4e34-8d8e-291c4c7f32a9",
  "rule": "475c31da-8b24-495e-aace-2ca8e57df622"
}
resp = requests.post(url, headers=h, data=json.dumps(data))
```

I believe this is correct from what I read.  @whitej6 @itdependsnetworks thoughts?

Reviewer: ok, perhaps that is correct, @whitej6 do you mind confirming?

Reviewer 2: It depends on the serializer I believe. If using jsonfield you shouldn't have to stringify the data in the payload

  "actual": "{\"aaa\": {\"servers\": [\"1.1.1.1\"]}}",
  "intended": "{\"aaa\": {\"servers\": [\"1.1.1.1\"]}}",
  "ordered": true,
  "compliance": true,
  "display": "nautobot-iosv-1 -> iosv-json - aaa-json -> True"
}
```

You can also see the compliance data in the UI once it is created via API.

In the navigation menu: `Plugins -> Configuration Compliance`.

![Example Compliance Run in UI](./img/03-navigating-compliance-json.png)

With the detailed view:

![Example Compliance Run in UI Detail](./img/04-navigating-compliance-json.png)

Example of a Non-Compliant rule:

![Example Non-Compliant Run in UI Detail](./img/05-navigating-compliance-json.png)

> Note:
All subsequent updates to the compliance rule need to use the `PUT` or `PATCH` API methods and they need to provide the UUID that was created for the specific compliance execution. This UUID is generated by (unique_together = ("device", "rule"))

As soon as a Configuration Compliance object has been created using the API seen above, the `Configuration Overview` will be available, and will show the **diff** action button.

![Example of Diff Icon](./img/06-navigating-compliance-json.png)

> This will **only** show up for devices using JSON compliance rules.

Reviewer: Can you clarify what "this" is?

> > Note: In the backend processing the plugin is finding all compliance rule objects that exist for the device and is creating the backup and intended configurations on the fly.  This all occurs when the **diff** action button is clicked.  When the modal is closed the Overview page is refreshed and the backup and intended configurations icons are now available.

Reviewer: ```suggestion
> Note: The backend will process and find all compliance rule objects that exist for the device and the create the a temporary backup and intended configurations on the fly. Meaning, when the **diff** action button is clicked all of the processes happens at that time, to render the diff.
```

> ![Example of Diff](./img/07-navigating-compliance-json.png)

Reviewer: can remove

> > Note: The backend will process and find all compliance rule objects that exist for the device and the create the a temporary backup and intended configurations on the fly. Meaning, when the **diff** action button is clicked all of the processes happens at that time, to render the diff.

Reviewer: Can you review this whole page, I don't think this is true either anymore.

Author: Got that fixed and fixed some of the screenshots.
